web/source/osk/visualKeyboard.ts

```diff
@@ -701,77 +697,29 @@ namespace com.keyman.osk {
         return;
       }
 
-      // Do not move over keys if device popup visible
-      if(this.popupVisible) {
-        if(key1 == null) {
-          if(key0) {
-            this.highlightKey(key0,false);
-          }
-          this.keyPending=null;
-          this.touchPending=null;
-        } else {
-          if(key1 == this.popupBaseKey) {
-            if(!key1.classList.contains('kmw-key-touched')) {
-              this.highlightKey(key1,true);
-            }
-            this.keyPending = key1;
-            this.touchPending = e.touches[0];
-          } else {
-            if(key0) {
-              this.highlightKey(key0,false);
-            }
-            this.keyPending = null;
-            this.touchPending = null;
-          }
-        }
+      // Update all gesture tracking.  The function returns true if further input processing
+      // should be blocked.
+      if(this.updateGestures(key1, key0, e.changedTouches[0])) {
         return;
       }
```

Comment:
Fortunately, after a lot of careful testing and analysis, it turns out that we can simply delegate all touch-handling to an active `RealizedGesture` whenever it exists.  If the gesture indicates that no further processing should occur, we're safe to exit without consequence.  If the `updateGestures` method notes that there is no `RealizedGesture`, it passes control back to the standard base-key handling.

This allows easy separation of subkey handling from base key handling, eliminating that nasty code spaghetti.